### PR TITLE
HZN-1518: Custom TTLs for detectors, monitors and collectors

### PR DESCRIPTION
**Pull request by cgorantla**
TTL specified in metadata overrides any other default TTL.
For Snmp (Collectors/NodeScan/Enlinkd)  TTL specified in agent configuration overrides metadata TTL or any other default TTL.

* JIRA: http://issues.opennms.org/browse/HZN-1518

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java [hunk 1]
```diff
@@ -48,12 +48,17 @@
...
+import org.springframework.transaction.support.TransactionOperations;
 
 import com.google.common.base.Strings;
 
 @Command(scope = "meta", name = "test", description = "Test Meta-Data replacement")
 @Service
 public class MetaCommand implements Action {
+
+    @Reference
+    private TransactionOperations transactionOperations;
```
Comment (j-white) [inline core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java]: Let's use `org.opennms.netmgt.dao.api.SessionUtils` instead.

File: core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java [hunk 2]
```diff
@@ -79,48 +84,53 @@ private void printScope(final Scope scope) {
         for (final Map.Entry<String, Set<ContextKey>> group : grouped.entrySet()) {
             System.out.printf("%s:\n", group.getKey());
             for (final ContextKey contextKey : group.getValue()) {
-                System.out.printf("  %s='%s'\n", contextKey.getKey(), scope.get(contextKey).get());
+                System.out.printf("  %s='%s'\n", contextKey.getKey(), scope.get(contextKey).isPresent() ? scope.get(contextKey).get() : "");
```
Comment (j-white) [inline core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java]: Can be simplified by using `.orElse("")`

File: core/snmp/api/src/main/java/org/opennms/netmgt/snmp/SnmpConfiguration.java [hunk 3]
```diff
@@ -232,6 +233,14 @@ public final void setWriteCommunity(String community) {
         m_writeCommunity = community;
     }
 
+    public Long getTTL() {
+        return m_ttl;
+    }
+
+    public void setTTL(Long ttl) {
+        this.m_ttl = m_ttl;
```
Comment (j-white) [inline core/snmp/api/src/main/java/org/opennms/netmgt/snmp/SnmpConfiguration.java]: `this` and `m_`

File: features/collection/client-rpc/src/main/java/org/opennms/netmgt/collection/client/rpc/CollectorRequestBuilderImpl.java [hunk 4]
```diff
@@ -150,6 +158,10 @@ public CollectorRequestBuilder withAttributes(Map<String, Object> attributes) {
         allAttributes.putAll(interpolatedAttributes);
         allAttributes.putAll(runtimeAttributes);
 
+        // For SNMP, ttl should be passed as parameter.
+        if(ttlInMs != null) {
+            allAttributes.put("SERVICE_INTERVAL", ttlInMs);
```
Comment (j-white) [inline features/collection/client-rpc/src/main/java/org/opennms/netmgt/collection/client/rpc/CollectorRequestBuilderImpl.java]: Can we find a better way to pass this in?

File: opennms-base-assembly/src/main/filtered/etc/poller-configuration.xml [hunk 5]
```diff
@@ -290,7 +290,7 @@
          <parameter key="period" value="30000"/><!-- Service interval should be same as period -->
       </service>
       <service name="Minion-RPC" interval="30000" user-defined="false" status="on">
-         <parameter key="ttl" value="30000"/><!-- TTL should be same as period -->
+         <parameter key="ttl" value="${service:ttl|interface:ttl|node:ttl|30000}"/><!-- TTL should be same as period -->
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/poller-configuration.xml]: Do we need to set this here? Also I think omitting the scope and just setting `${ttl|30000}` would be equivalent.

File: opennms-provision/opennms-detectorclient-rpc/src/main/java/org/opennms/netmgt/provision/detector/client/rpc/DetectorRequestBuilderImpl.java [hunk 6]
```diff
@@ -50,6 +52,8 @@
 
     private static final String PORT = "port";
 
+    private static final Long  DEFAULT_VALUE = -1L;
```
Comment (j-white) [inline opennms-provision/opennms-detectorclient-rpc/src/main/java/org/opennms/netmgt/provision/detector/client/rpc/DetectorRequestBuilderImpl.java]: Unused?

File: core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java [hunk 7]
```diff
@@ -79,48 +84,53 @@ private void printScope(final Scope scope) {
...
     public Object execute() throws Exception {
+
+        transactionOperations.execute(status -> {
         try {
-            final OnmsNode onmsNode = this.nodeDao.get(this.node);
-            if (onmsNode == null) {
-                System.out.printf("Cannot find node with ID/FS:FID=%s.\n", this.node);
+                final OnmsNode onmsNode = this.nodeDao.get(this.node);
+                if (onmsNode == null) {
+                    System.out.printf("Cannot find node with ID/FS:FID=%s.\n", this.node);
```
Comment (mvrueden) [inline core/ipc/rpc/shell-commands/src/main/java/org/opennms/core/rpc/commands/MetaCommand.java]: This is misleading. `onmsNode` is the node object, where `this.node` is either a node id or a `FS:FID`. In the graph world I came up with the term `nodeRef` or `nodeReference` which can either be the id or `FS:FID`. I would suggest using that terminology here as well, as it makes things much more clear.

**Comment by cgorantla** [comment]
@j-white @mvrueden  Handled review comments.  Please take a look once again.

**Pull request merged**